non-blocking-da v1: da errgroup

samlaf opened the pull request:
This is the simplest possible implementation to fix the non-blocking da requests. We build off of @karlb's https://github.com/celo-org/optimism/pull/213. We use an errgroup for da requests, with its own limit to the number of concurrent goroutines. See https://hackmd.io/@samlaf/op-batcher-concurrent-altda-requests-design-doc for more details.

### Testing
We added a basic [e2e test](https://github.com/Layr-Labs/optimism/pull/7/files#diff-8bc46de6e90dc6a2a1773e2a90eca15e631d5b7e062fc2f282371082faa30e52R1406) by using a fakeDaServer that sleeps for 5 seconds before returning commitments, in order to simulate slow EigenDA responses.

We also tested this manually by running the devnet and submitting large txs and observing the behavior.

samlaf committed:
feat: initial goroutine blob submission implementation

test(batcher): add e2e test for concurrent altda requests

doc: add explanation comment for FakeDAServer

chore: fix if condition in altda sendTransaction path

feat: add maxConcurrentDaRequests config flag + semaphore

refactor: batcher to use errgroup for da instead of separate semaphore/waitgroup

fix: nil pointer bug after using wrong function after rebase

fix: defn of maxConcurrentDaRequests=0

fix: TestBatcherConcurrentAltDARequests

chore: remove unneeded if statement around time.Sleep

refactor: use TryGo instead of Go to make logic local and easier to read

chore: clean up some comments in batcher

chore: make batcher shutdown cancel pending altda requests by using shutdownCtx instead of killCtx

Inphi committed:
cannon: Fix stack patching (#11632)

* cannon: Fix stack patching

And add `memprofilerate=0` to envp

* Update cannon/mipsevm/program/patch.go

Co-authored-by: protolambda <proto@protolambda.com>

* cleanup argv/envp string ptrs

* nit

* fix envar name

* Update cannon/mipsevm/program/patch.go

Co-authored-by: mbaxter <meredith@oplabs.co>

* align op-program arg0

---------

Co-authored-by: protolambda <proto@protolambda.com>
Co-authored-by: mbaxter <meredith@oplabs.co>

mslipper committed:
Archive contract artifacts (#11626)

To demonstrate how we can make our deployments more modular, this PR proposes archiving smart contract artifacts as tarballs that get uploaded to GCS. This  allows deployment tools to use precompiled artifacts rather than generating them on-the fly.

The archives are named after the hash of all Solidity files in the contracts-bedrock folder, including those in `lib/`, plus some additional metadata files like `foundry.toml` and `semver-lock.json`. See `calculate-checksum.sh` for details on how the algorithm works. I'm open to feedback around what should make up the checksum.

Since the atifacts are content-addressable, this PR also updates the CI pipeline to download the artifacts from GCS prior to running `pnpm monorepo`. When the Solidity codebase doesn't change, this allows the `pnpm monorepo` job to skip compiling Solidity altogether. While this won't work as well when we're actively modifying the Solidity codebase, since the hash will change, it does provide a modest speedup in CI.

mbaxter committed:
cannon: Add MTCannon-specific differential tests (#11605)

* cannon: Implement multithreaded clone fuzz test

* cannon: Add more clone evm tests

* cannon: Add evm test for GetTID syscall

* cannon: Add evm test for SysExit

* cannon: Add evm test for popping exited threads from the stack

* cannon: Fix futex wait handling, add evm test

* cannon: Add evm test for handling waiting thread

* cannon: Add test utils for defining / validating MTState expectations

* cannon: Add tests for futex wake, wake traversal

* cannon: Add test for SysYield

* cannon: Add SysOpen test, todos

* cannon: Add test for SchedQuantum preemption, fix inconsistency

* cannon: Add tests for noop, unsupported syscalls

* cannon: Remove duplicate constants

* cannon: Add tests for unsupported futex ops

* cannon: Group traversal tests, fix TestEVM_WakeupTraversalStep

* cannon: Add tests for nanosleep

* cannon: Add additional testcase for wakeup traversal

* cannon: Tweak futex wake tests

* cannon: Update mt fuzz test to use new test utils

* cannon: Rename contructor method for consistency

* cannon: Add some simple tests for ExpectedMTState util

* cannon: Add another validation test

* cannon: Move syscall lists to tests where they're used

* cannon: Add comment

* cannon: Extract some evm test helpers

* cannon: Cleanup - use require.Equalf for formatting

* cannon: Rename test util to AssertEVMReverts

* cannon: Add GetThreadStacks helper

* cannon: Add a few more traversal tests

anacrolix committed:
Conductor and sequencer p2p refactoring (#11455)

* Shutdown sequencer before stopping p2p

* Check p2p isn't also disabled

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

* Remove missed time.Sleep

* Fix up use of SetupP2P.Disabled

* Revert error check after RPC boundary

* Add comment about context for StopSequencer

* Add Config.p2pEnabled

* op-node: Make Config.P2PEnabled public

---------

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

agusduha committed:
test: fix standard bridge interop tests (#11668)

* test: fix L2 standard bridge interop tests

* test: mock factory implementation instead of proxy

Commit:
- make state publishing loop abort if the txpool state is not good (#11633)

- protect txpool state vars with a mutex so they can be automically updated to avoid potential race
  condition

dependabot[bot] committed:
dependabot(gomod): bump github.com/hashicorp/raft from 1.7.0 to 1.7.1 (#11670)

Bumps [github.com/hashicorp/raft](https://github.com/hashicorp/raft) from 1.7.0 to 1.7.1.
- [Release notes](https://github.com/hashicorp/raft/releases)
- [Changelog](https://github.com/hashicorp/raft/blob/main/CHANGELOG.md)
- [Commits](https://github.com/hashicorp/raft/compare/v1.7.0...v1.7.1)

---
updated-dependencies:
- dependency-name: github.com/hashicorp/raft
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bxue-l2 commented on op-batcher/batcher/driver.go:
```diff
@@ -483,7 +489,13 @@ func (l *BatchSubmitter) clearState(ctx context.Context) {
+func (l *BatchSubmitter) publishTxToL1(ctx context.Context, queue *txmgr.Queue[txRef], receiptsCh chan txmgr.TxReceipt[txRef], daWaitGroup *sync.WaitGroup, daSemaphore *semaphore.Weighted) error {
+	// Acquire semaphore or return immediately if it is at capacity
+	// we don't want to pull data out of the state if we won't be able to send it
+	if l.Config.UseAltDA && !daSemaphore.TryAcquire(1) {
```
should we overload error, or just create a new error. Since there is only one caller function, I will prefer new error

bxue-l2 commented on op-batcher/batcher/driver.go:
```diff
@@ -550,9 +562,40 @@ func (l *BatchSubmitter) cancelBlockingTx(queue *txmgr.Queue[txRef], receiptsCh
+		daWaitGroup.Add(1)
+		go func() {
+			defer daWaitGroup.Done()
+			defer daSemaphore.Release(1)
```
defer is last-in first-out. Should the order be switched? first say it is done, then release

samlaf replied:
https://cs.opensource.google/go/x/sync/+/refs/tags/v0.8.0:errgroup/errgroup.go;l=38
Seems like errgroup also releases the sem before noticing the waitgroup.
Not sure if either matters... but I guess we should do whatever errgroup is doing? I can add a comment

samlaf replied:
actually switched to an errgroup.

anacrolix committed:
Use context interrupts consistently in more places (#11511)

* Use context interrupts consistently in more places

* Fix CI lint errors

(cherry picked from commit 0410b7e448e063fa9e30295fbe423ff1c0171d12)

* op-service/ctxinterrupt: address review comments

---------

Co-authored-by: protolambda <proto@protolambda.com>

karlb commented on op-alt-da/damock.go:
```diff
@@ -99,3 +103,88 @@ func (d *AltDADisabled) OnFinalizedHeadSignal(f HeadSignalFn) {
+}
+
+func (s *FakeDAServer) HandleGet(w http.ResponseWriter, r *http.Request) {
+	if s.getRequestLatency > 0 {
+		time.Sleep(s.getRequestLatency)
+	}
```
We could do without the `if`, since numbers <=0 will make sleep return immediately.

samlaf replied:
Unfortunately not. I wrote the explanation in the comment right above. Only < 0 means no limit. 0 literally means no goroutine can start. See https://github.com/golang/go/issues/27837#issuecomment-1121177106

karlb replied:
Are you mixing this up with `MaxConcurrentDARequests`? This is just a sleep, no goroutines involved.

samlaf replied:
I was! shouldn't review PRs on my phone haha

samlaf replied:
Fixed: [dbcec39](https://github.com/Layr-Labs/optimism/pull/7/commits/dbcec398525377cb9f212c6ed8b8e3f5abde4695)

karlb commented on op-batcher/batcher/driver.go:
```diff
@@ -550,9 +567,39 @@ func (l *BatchSubmitter) cancelBlockingTx(queue *txmgr.Queue[txRef], receiptsCh
+		// when posting txdata to an external DA Provider, we use a goroutine to avoid blocking the main loop
+		// since it may take a while for the request to return.
+		// the number of concurrent goroutines is limited by the semaphore check (daGroup.TryAcquire(1)) in publishTxToL1
+		daGroup.Go(func() error {
```
At a first glance, this looks like it might block due to the errgroup limit. A goroutine could be started by a different process between the `tryGo` in `publishTxToL1` and here.

But seeing the whole program, we know that both are only called in the main loop, so we can actually do this safely. And it is simpler than the concurrency-safe approach.

samlaf replied:
The concurrency-safe approach being to use TryGo again here and put back the frames if didn’t start? Honestly the more I think about it the more I don’t dislike that approach. Also means we can get rid of the previous tryGo and we make the logic more localized, as you are explaining. Probably leads to less confusion?

karlb replied:
> The concurrency-safe approach being to use TryGo again here and put back the frames if didn’t start?

Either this, or the fiddly acquire early, start goroutine later approach.

> Honestly the more I think about it the more I don’t dislike that approach. Also means we can get rid of the previous tryGo and we make the logic more localized, as you are explaining. Probably leads to less confusion?

I also think it is clearer and gets rid of the strange TryGo that doesn't actually want to do anything. As long as there is not serious performance impact (and I don't think it is) that is probably the better approach.

samlaf replied:
Thanks for that PR! Incorporated it with some other small modifs about logging that we talked about during our sync: [077204d](https://github.com/Layr-Labs/optimism/pull/7/commits/077204d7acb084b218881f34b74d32f61a5b4558)

karlb reviewed (commented):
Looks good! I'll give it a second read through later today.

agusduha committed:
test: fix superchain erc20 invariants (#11688)

---------

Co-authored-by: 0xDiscotech <131301107+0xDiscotech@users.noreply.github.com>

agusduha committed:
feat: add createX preinstall (#29) (#11618)

* feat: add createX preinstall

* feat: change name from CreateXDeployer to CreateX

mslipper committed:
Update contracts publish job to use env vars rather than contexts (#11687)

Contexts are scoped to a specific GitHub user group, which doesn't work with the GitHub merge queue or OSS contributors. This PR updates the packaging job to use raw project-level env vars instead (which are not user-scoped), and to only run on commits to `develop`.

samlaf committed:
chore(batcher): make altda wg wait + log only when useAltDa is true

samlaf committed:
refactor: batcher altda submission code into its own function

samlaf committed:
test: refactor batcher e2e test to only count batcher txs

samlaf committed:
chore: log errors from wait functions

samlaf committed:
chore: refactor and minimize time that e2e batcher system tests can run

samlaf committed:
chore: lower timeout duration in test

samlaf committed:
fix(batcher): maxConcurentDARequests was not being initialized

samlaf commented:
Closing since this was merged upstream.
